spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/CassandraSourceRelation.scala

```diff
@@ -0,0 +1,201 @@
+            userSpecifiedSchema: Option[StructType], sqlContext: SQLContext) = scanType.makeRelation(
+    table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Base table scan relation*/
+case object BaseScanType extends ScanType  {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new TableScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns table scan relation*/
+case object PrunedScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new PrunedScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and filtered table scan relation*/
+case object PrunedFilteredScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new PrunedFilteredScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Pruned columns and Catalyst filtered table scan relation*/
+case object CatalystScanType extends ScanType {
+  override def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                            userSpecifiedSchema: Option[StructType],sqlContext: SQLContext) =
+    new CatalystScanRelationImpl(table, keyspace, cluster, userSpecifiedSchema, sqlContext)
+}
+
+/** Table scan type */
+sealed trait ScanType {
+  def makeRelation(table: String, keyspace: String, cluster: Option[String],
+                   userSpecifiedSchema: Option[StructType], sqlContext: SQLContext) : BaseRelationImpl
+}
+
+/** Base relation implements [[BaseRelation]] and [[InsertableRelation]] */
+private[cassandra] class BaseRelationImpl(table: String, keyspace: String, cluster: Option[String],
+                                          userSpecifiedSchema: Option[StructType], sqlcontext: SQLContext)
+  extends BaseRelation with InsertableRelation with Serializable with Logging {
+
+  override def sqlContext: SQLContext = sqlcontext
+
+  protected[this] val tableDef = sqlContext
+    .schemas.get(cluster.getOrElse("default"))
+    .keyspaceByName.getOrElse(keyspace, throw new IOException(s"Keyspace not found: $keyspace"))
+    .tableByName.getOrElse(table, throw new IOException(s"Table not found: $keyspace.$table"))
+
+  protected[this] val columnNameByLowercase = tableDef.allColumns
+    .map(c => (c.columnName.toLowerCase, c.columnName)).toMap
+
+  override def schema: StructType = {
+    def columnToStructField(column: ColumnDef): StructField = {
+      StructField(column.columnName, ColumnDataType.catalystDataType(column.columnType, nullable = true))
```

Comment:
What does it mean exactly - why columns are nullable? AFAIK Connector doesn't support nulls.